File: src/main/java/org/organicdesign/fp/collections/UnmodListIterator.java
package org.organicdesign.fp.collections;

import java.util.ListIterator;

/** An unmodifiable ListIterator */
public interface UnmodListIterator<E> extends ListIterator<E>, UnmodSortedIterator<E> {

    // ========================================= Instance =========================================
    /** Not allowed - this is supposed to be unmodifiable */
    @Override @Deprecated default void add(E element) {
        throw new UnsupportedOperationException("Modification attempted");
    }

// boolean	hasNext()
// boolean	hasPrevious()
// E	next()
// int	nextIndex()
// E	previous()

    // I think this is the only valid implementation of this method. You can override it if you
    // think otherwise.
    /** {@inheritDoc} */
    @Override default int previousIndex() { return nextIndex() - 1; }

    /** Not allowed - this is supposed to be unmodifiable */
    @SuppressWarnings("deprecation")
    @Override @Deprecated default void remove() {
        throw new UnsupportedOperationException("Modification attempted");
    }

    /** Not allowed - this is supposed to be unmodifiable */
    @Override @Deprecated default void set(E element) {
        throw new UnsupportedOperationException("Modification attempted");
    }

// Methods inherited from interface java.util.Iterator
// forEachRemaining
}

File: src/main/java/org/organicdesign/fp/collections/UnmodListIterator.kt
package org.organicdesign.fp.collections

/** An unmodifiable ListIterator  */
//@PurelyImplements("java.util.ListIterator")
interface UnmodListIterator<E> : ListIterator<E>, UnmodSortedIterator<E> {


    // boolean	hasNext()
    // boolean	hasPrevious()
    // E	next()
    // int	nextIndex()
    // E	previous()

    // I think this is the only valid implementation of this method. You can override it if you
    // think otherwise.
    @JvmDefault
    override fun previousIndex(): Int = nextIndex() - 1

    // Methods inherited from interface java.util.Iterator
    // forEachRemaining
}
